### PR TITLE
Defensive Editor Diagnostics

### PR DESCRIPTION
**Pull request by RobertBColton**
This pull request is the product of the discussion in #62. There is a mutual consensus that we should use defensive programming conventions in RGM. This includes logging warnings in release mode and debug mode. Here I have added such defensive logging for model indexes provided to most of our model classes.

The `qWarning` macro was used for logging with no use of assertions. I don't see it as necessary to force the program to exit when it is capable of recovering, and adding all of these as assertions too will just make this diff even fatter. We have made the decision to reserve these Qt macros for internal RGM issues only. If we later decide to reuse QDebug for controlling the build output verbosity, then we plan to use a special instance of the QDebug class.

Me and Josh collaborated on `Components/Logger.h` to make this schizophrenic precondition checking less redundant. It's actually becoming more standard practice in modern years, see Mozilla Firefox.

### Toggle Editor Diagnostics
Toggle Editor Diagnostics
### Example Editor Diagnostics
Example Editor Diagnostics

### CODE REVIEW

File: Models/ProtoModel.cpp [hunk 1]
```diff
@@ -128,6 +135,10 @@ QVariant ProtoModel::data(int row, int column) const {
 }
 
 QVariant ProtoModel::data(const QModelIndex &index, int role) const {
+  if (!index.isValid()) {
+    qWarning() << index << "provided to ProtoModel::data for role" << role << "is invalid.";
```
Comment (JoshDreamland) [inline Models/ProtoModel.cpp]: I love how every strange, far-fetched thing I tell you to do, Qt has a builtin for.

Comment (RobertBColton) [reply Models/ProtoModel.cpp]: Actually that's kind of funny now that you mention it. Kind of a bit of a bitter sweet irony I'd say.

Comment (JoshDreamland) [reply Models/ProtoModel.cpp]: No, it just means that Qt is keeping up with the trends. This sort of logger is becoming prevalent, and it's nice to see it included in the framework. It makes a ton of sense for the UI framework to just handle this, and it saves us a bunch of code. All while still giving us access to the information I wanted (eg, source file + line number). It's great.

Comment (RobertBColton) [reply Models/ProtoModel.cpp]: Sure, I get that, I just thought you were being tongue in cheek too. But ok, great!

File: Models/ProtoModel.cpp [hunk 2]
```diff
@@ -78,6 +81,10 @@ bool ProtoModel::IsDirty() { return dirty; }
 int ProtoModel::columnCount(const QModelIndex & /*parent*/) const { return 1; }
 
 bool ProtoModel::setData(const QModelIndex &index, const QVariant &value, int role) {
+  if (!index.isValid()) {
+    qWarning() << index << "provided to ProtoModel::setData for role" << role << "with value" << value << "is invalid.";
```
Comment (JoshDreamland) [inline Models/ProtoModel.cpp]: You don't have spaces around those values. Or any others. How can Qt know to separate them? And how does it delimit them?

Comment (RobertBColton) [reply Models/ProtoModel.cpp]: It's built into `qDebug`, which `qWarning` macro uses.
https://doc.qt.io/qt-5/qdebug.html#basic-use

File: Components/Logger.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+#include <QDebug>
+#include <QtGlobal>
+
+class LogAndReturn {
+  QDebug logger;
+
+ public:
+  LogAndReturn(QtMsgType msgType) : logger(msgType) {}
+  LogAndReturn(const QDebug& log) : logger(log) {}
```
Comment (JoshDreamland) [inline Components/Logger.h]: QDebug may implement move semantics (ie, you may be able to construct using `QDebug&&` from that `qWarning()` object.

Comment (RobertBColton) [reply Components/Logger.h]: Ok: d99691125dcb9f17107fdca705b98c43cd4192fd

Qt seems cool with that.

Comment (RobertBColton) [reply Components/Logger.h]: Also here's the definition of `qWarning()`.

```cpp
#define qWarning QMessageLogger(QT_MESSAGELOG_FILE, QT_MESSAGELOG_LINE, QT_MESSAGELOG_FUNC).warning
```

**Pull request merged**